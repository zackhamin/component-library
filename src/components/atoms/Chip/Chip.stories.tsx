import { Check } from 'Atoms/Icons';
import { Story } from '@storybook/react';

import { Chip, PolymorphicChipProps } from './Chip';

export default {
	title: 'atoms/Chip',
	argTypes: {
		children: {
			description: 'text displayed within the chip',
			control: {
				type: 'text',
			},
		},
		as: {
			control: {
				type: 'select',
				options: ['a', 'button'],
			},
		},
		icon: {
			control: {
				type: 'select',
				options: ['CHECK', 'CHEVRON', 'ERROR'],
			},
		},
		variant: {
			control: {
				type: 'select',
				options: ['default', 'topic'],
			},
		},
	},
};

const Template: Story<PolymorphicChipProps<'button'>> = (args) => (
	<Chip {...args} />
);
export const Default = Template.bind({});
Default.args = {
	children: '8:00',
};

export const DefaultIcon = Template.bind({});
DefaultIcon.args = {
	icon: <Check />,
	children: '8:00',
};

export const Topic = Template.bind({});
Topic.args = {
	variant: 'topic',
	children: 'Clean Beauty',
};

export const TopicIcon = Template.bind({});
TopicIcon.args = {
	variant: 'topic',
	icon: <Check />,
	children: 'Clean Beauty',
};
