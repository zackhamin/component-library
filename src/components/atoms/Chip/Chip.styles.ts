import { ReactNode } from 'react';
import styled from 'styled-components';

import { colors, typography } from 'Styles';

const StyledChip = styled.button<{
	icon?: ReactNode;
}>`
	background-color: white;
	border: none;
	transition: background-color 0.15s ease-in-out;
	font-family: ${typography.type.heading};
	font-size: 1.8rem;
	cursor: pointer;
	display: inline-flex;
	align-items: center;

	&:focus-visible {
		outline: solid 2px ${colors.emerald};
		outline-offset: 4px;
	}

	&.variant-default {
		padding: ${(p) => (p.icon ? '1.6rem 2.65rem' : '0.8rem')};
		box-shadow: 0px 2px 6px rgba(51, 51, 51, 0.16);
		border-radius: 0.8rem;
		color: ${colors.emerald};

		.icon-wrapper {
			display: flex;
			margin-right: 1rem;
		}

		&:hover {
			background-color: ${colors.grey300};
			color: ${colors.emerald};
			text-decoration: none;
		}
	}

	&.variant-topic {
		padding: ${(p) =>
			p.icon ? '1.2rem 1.8rem 1.2rem 1.2rem' : '1.2rem 1.6rem'};
		border-radius: 3rem;
		background-color: ${colors.grey200};
		color: ${colors.offBlack};

		.icon-wrapper {
			width: 3.6rem;
			height: 3.6rem;
			display: flex;
			align-items: center;
			justify-content: center;
			border-radius: 50%;
			background-color: ${colors.emerald25};
			margin-right: 0.6rem;
		}

		&:hover {
			color: ${colors.offBlack};
			text-decoration: none;
		}
	}
`;

export { StyledChip };
