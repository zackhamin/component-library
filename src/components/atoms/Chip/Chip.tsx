import { StyledChip } from './Chip.styles';

export interface ChipProps<Component extends React.ElementType> {
	as?: Component;
	children: React.ReactNode;
	icon?: React.ReactNode;
	variant?: 'default' | 'topic';
	ariaLabel?: string;
}

// https://www.benmvp.com/blog/polymorphic-react-components-typescript/
export type PolymorphicChipProps<E extends React.ElementType> = ChipProps<E> &
	Omit<React.ComponentProps<E>, keyof ChipProps<E>>;

const Chip = <Component extends React.ElementType = 'button'>({
	as,
	children,
	icon,
	ariaLabel,
	variant = 'default',
	...rest
}: PolymorphicChipProps<Component>) => {
	const tag = as || 'button';

	return (
		<StyledChip
			icon={icon}
			aria-label={ariaLabel}
			className={`variant-${variant}`}
			as={tag}
			{...rest}
		>
			{icon && <div className="icon-wrapper">{icon}</div>}
			{children}
		</StyledChip>
	);
};

export { Chip };
