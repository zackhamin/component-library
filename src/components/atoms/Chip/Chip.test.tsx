import { Chip } from '../Chip';
import { render, screen } from '@testing-library/react';

describe('Chip', () => {
	it('renders correctly', () => {
		const { container } = render(<Chip>Chip</Chip>);
		expect(container).toMatchSnapshot();
	});

	it('correctly renders default variant type', () => {
		render(<Chip>Chip</Chip>);

		expect(screen.getByRole('button')).toHaveClass('variant-default');
	});

	it('correctly renders topic variant type', () => {
		render(<Chip variant="topic">Chip</Chip>);

		expect(screen.getByRole('button')).toHaveClass('variant-topic');
	});

	it('correctly handles polymorphism', () => {
		render(
			<Chip as="a" href="/" variant="topic">
				Chip
			</Chip>
		);

		expect(screen.getByRole('link')).toBeInTheDocument();
	});
});
