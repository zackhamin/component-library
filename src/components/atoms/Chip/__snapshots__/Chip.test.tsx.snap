// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip renders correctly 1`] = `
.c0 {
  background-color: white;
  border: none;
  -webkit-transition: background-color 0.15s ease-in-out;
  transition: background-color 0.15s ease-in-out;
  font-family: BrandonGrotesque-Bold,Arial,Helvetica,sans-serif;
  font-size: 1.8rem;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:focus-visible {
  outline: solid 2px #00594C;
  outline-offset: 4px;
}

.c0.variant-default {
  padding: 0.8rem;
  box-shadow: 0px 2px 6px rgba(51,51,51,0.16);
  border-radius: 0.8rem;
  color: #00594C;
}

.c0.variant-default .icon-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 1rem;
}

.c0.variant-default:hover {
  background-color: #eff5f3;
  color: #00594C;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0.variant-topic {
  padding: 1.2rem 1.6rem;
  border-radius: 3rem;
  background-color: #F6F6F6;
  color: #333333;
}

.c0.variant-topic .icon-wrapper {
  width: 3.6rem;
  height: 3.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 50%;
  background-color: #BFE0DC;
  margin-right: 0.6rem;
}

.c0.variant-topic:hover {
  color: #333333;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div>
  <button
    class="c0 variant-default"
  >
    Chip
  </button>
</div>
`;
